gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,792 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from gensim import utils
+import logging
+from timeit import default_timer
+import threading
+import numpy as np
+from six import string_types
+from six.moves import xrange
+from numpy import vstack
+from gensim import matutils
+from numpy import float32 as REAL, ones, random, argmax
+from types import GeneratorType
+from gensim.utils import deprecated
+
+try:
+    from queue import Queue
+except ImportError:
+    from Queue import Queue
+
+logger = logging.getLogger(__name__)
+
+
+# Public Interfaces
+class BaseAny2VecModel(utils.SaveLoad):
+
+    def __init__(self, workers=3, vector_size=100, epochs=5,
+                 callbacks=(), batch_words=10000):
+        """Initialize model parameters.
+        Subclass should initialize the following attributes:
+        - self.kv (instance of concrete implementation of `BaseKeyedVectors` interface)
+        - self.vocabulary (instance of concrete implementation of `BaseVocabBuilder` abstract class)
+        - self.trainables (instance of concrete implementation of `BaseTrainables` abstract clas)
+        """
+        self.vector_size = int(vector_size)
+        self.workers = int(workers)
+        self.epochs = epochs
+        self.train_count = 0
+        self.total_train_time = 0
+        self.batch_words = batch_words
+        self.model_trimmed_post_training = False
+        self.callbacks = callbacks
+
+    def build_vocab(self, data_iterable, update=False, progress_per=10000, **kwargs):
+        """Scan through all the data and create/update vocabulary.
+        Should also initialize/reset/update vectors for new vocab entities.
+        """
+        raise NotImplementedError
```

Comment:
Better raise instance, not a class (tip from https://www.python.org/dev/peps/pep-0008/), here and everywhere.